src/core/qgsuserprofilemanager.cpp

```diff
@@ -0,0 +1,200 @@
+#include "qgsuserprofilemanager.h"
+#include "qgsuserprofile.h"
+#include "qgsapplication.h"
+#include "qgslogger.h"
+
+#include <QFile>
+#include <QDir>
+#include <QTextStream>
+#include <QProcess>
+#include <QStandardPaths>
+
+
+QgsUserProfileManager::QgsUserProfileManager( const QString &rootLocation, QObject *parent )
+  : QObject( parent )
+{
+  mWatcher.reset( new QFileSystemWatcher() );
+  setRootLocation( rootLocation );
+  connect( mWatcher.get(), &QFileSystemWatcher::directoryChanged, this, [this]
+  {
+    emit profilesChanged();
+  } );
+}
+
+QString QgsUserProfileManager::resolveProfilesFolder( const QString &basePath )
+{
+  if ( getenv( "QGIS_CUSTOM_CONFIG_PATH" ) )
```

Comment:
As suggested in the general comments, I'd prefer if the `getenv` would be moved to main.cpp and its value passed as an argument to the constructor as an input to this class.